src/source.rs

```diff
@@ -0,0 +1,347 @@
+use std::{
+  borrow::Cow,
+  convert::{TryFrom, TryInto},
+  fmt,
+  hash::{Hash, Hasher},
+};
+
+use serde::{Deserialize, Serialize};
+
+use crate::{
+  helpers::{decode_mappings, StreamChunks},
+  Result,
+};
+
+pub type BoxSource = Box<dyn Source>;
+
+pub trait Source: StreamChunks + DynHash + fmt::Debug + Sync + Send {
+  fn source(&self) -> Cow<str>;
+
+  fn buffer(&self) -> Cow<[u8]>;
+
+  fn size(&self) -> usize;
+
+  fn map(&self, options: &MapOptions) -> Option<SourceMap>;
+
+  fn update_hash(&self, state: &mut dyn Hasher) {
+    self.dyn_hash(state);
+  }
+}
+
+impl<T: Source + 'static> From<T> for BoxSource {
+  fn from(s: T) -> Self {
+    Box::new(s)
+  }
+}
+
+// for `updateHash`
+pub trait DynHash {
+  fn dyn_hash(&self, state: &mut dyn Hasher);
+}
+
+impl<H: Hash> DynHash for H {
+  fn dyn_hash(&self, mut state: &mut dyn Hasher) {
+    self.hash(&mut state);
+  }
+}
+
+impl Hash for dyn Source {
+  fn hash<H: Hasher>(&self, state: &mut H) {
+    self.dyn_hash(state);
+  }
+}
+
+pub trait SourceExt {
+  fn boxed(self) -> BoxSource;
+}
+
+impl<T: Source + 'static> SourceExt for T {
```

Comment:
Is this necessary? Since we already have a fast-pass for converting `Source` to `BoxSource`

Reply:
```rust
new ConcatSource([
  Box::new(RawSource("aaa")) as BoxSource,
  Box::new(OriginalSource("bbb")),
])
```

vs.

```rust
new ConcatSource([
  RawSource("aaa")).boxed(),
  OriginalSource("bbb").boxed(),
])
```

only for less typing, I can delete it if you think it isn't necessary.

Reply:
You are right, it will be useful if multiple `Source` implementations are provided.